lib/posix/pthread.c

```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
 		return ESRCH;
 	}
 
-	pthread_mutex_lock(&pthread->state_lock);
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (pthread->state == PTHREAD_DETACHED) {
```

Comment:
This is fetched outside of the lock; I think this only assumes state can be read atomically, right? Threads can't move from DETATCHED to some state other than TERMINATED?

Reply:
Yeah, i noticed that too. @andyross - any thoughts here?

Reply:
State is a word, so in practice it's atomic.  But yeah, outside the lock that gets squishy and we should tighten it up, which is a little complicated given that these states are just recapitulating the (separately synchronized!) kernel internal state.  Maybe a separate boolean/flag for "detached" would be easier to manage, since it can change in only one direction.

Not sure how that race can get you back to the scheduler crash, though.  That should only happen if you have a k_thread_create() call that begins before some context returns from k_thread_abort() or k_thread_join().  Maybe the test has a state where it ignores the error from pthread_join() (where it fails to join a detached thread) and is trying to recreate it before it exits?  I don't think pthreads actually gives you an API to handle that safely at all, since you can't use join...

Reply:
Used a local bool.